src/Interpreters/Context.cpp

```diff
@@ -1052,25 +1052,50 @@ ConfigurationPtr Context::getUsersConfig()
     return shared->users_config;
 }
 
-void Context::setUser(const UUID & user_id_)
+void Context::setUser(const UUID & user_id_, bool set_current_profiles_, bool set_current_roles_, bool set_current_database_)
 {
-    auto lock = getLock();
+    /// Prepare lists of user's profiles, constraints, settings, roles.
 
-    user_id = user_id_;
+    std::shared_ptr<const User> user;
+    std::shared_ptr<const ContextAccess> temp_access;
+    if (set_current_profiles_ || set_current_roles_ || set_current_database_)
+    {
+        /// `temp_access` is used here only to extract information about the user, not to actually check access.
+        /// NOTE: AccessControl::getContextAccess() may require some IO work, so Context::getLock() must be unlocked while we're doing this.
+        ContextAccessParams params{user_id, /* full_access= */ false, /* use_default_roles = */ true, {}, {}, {}, {}};
+        temp_access = getAccessControl().getContextAccess(params);
```

Comment:
The issue existed because `AccessControl::getContextAccess()` may decide to do some IO work (for example, it can load the definition of a used role from disk). That IO work can fail in another thread and try to write information about the failure to `ProcessList` which could require access to the context again but from the other thread.